### DISCUSSION TITLE
Controlling valves outlets using Pyswmm

### DISCUSSION THREAD

**Discussion by Dinho4** [pyswmm/pyswmm#330] [Q&A] [answered]
The following error occurs when I try to open valves:
>>> from pyswmm import Simulation, Links
>>> for orif in orifices:
...     orif.target_setting = 1.0
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'orifices' is not defined

**Comment by jennwuu** [comment]
Hey @Dinho4 ,

You need to define orifices object first in Python. Please check out pyswmm documentation for how to access link parameters. 
https://pyswmm.readthedocs.io/en/latest/tutorial/tutorial.html#links

**Comment by Dinho4** [comment]
The following occurs after I followed the documentation:
>>> for orif in orifices:
...     orif.target_setting = 1.0
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'Link' object is not iterable

**Reply by jennwuu** [reply]
Please share the entire code snippet. 

Remember to import the modules from pyswmm.

```
from pyswmm import Links, Nodes, Simulation
```

**Comment by bemcdonnell** [comment, marked as answer]
```python

from pyswmm import Simulation, Links

with Simulation("Input_file.inp") as sim:
    all_links = Links(sim)

    # Use a list comprehension to grab only the orifices. 
    orifices = [link for link in all_links if link.is_orifice()]

    # Start your simulation
    for ind, step in enumerate(sim):
        for orifice in orifices:
            orifice.target_setting = 1 # or whatever number here. 

```

**Reply by Dinho4** [reply]
Thanks @bemcdonnell and @jennwuu, I was able to define orifices object first in Python.
